File: control/state.lua
---@param surface LuaSurface
local function setNextWarpzone(surface)
    storage.next_warp_zone = storage.next_warp_zone + 1
    storage.start_tick = game.tick
    storage.current_surface = surface;
end

local function nextWarpzone()
    return storage.next_warp_zone
end

local function currentWarpzone()
    return nextWarpzone() - 1
end

local function getWarpzoneTicks()
    return game.tick - storage.start_tick
end

local function surfaces()
    return {
        factory = game.surfaces.nauvis,
        ground = storage.current_surface
    }
end

local function nextWarpzoneSeed()
    return storage.generator()
end

local function onInit()
    storage.next_warp_zone = 1
    storage.start_tick = 0
    storage.generator = game.create_random_generator()
end

return { surfaces = surfaces, getWarpzoneTicks = getWarpzoneTicks, setNextWarpzone = setNextWarpzone, onInit = onInit, currentWarpzone = currentWarpzone, nextWarpzone = nextWarpzone, nextWarpzoneSeed = nextWarpzoneSeed }

File: control/state.ts
import { surfaces } from "constants";
import { LuaEntity, LuaSurface, OnObjectDestroyedEvent } from "factorio:runtime";

declare const storage: {
    next_warp_zone: number;
    warp_zone_start_tick: number;
    current_surface: LuaSurface;
    reactor: LuaEntity;
    reactor_destruction: number;
};

export function setNextWarpzone(surface: LuaSurface) {
    storage.next_warp_zone = storage.next_warp_zone + 1;
    storage.warp_zone_start_tick = game.tick;
    storage.current_surface = surface;
}

export function nextWarpzone() {
    return storage.next_warp_zone;
}

export function currentWarpzone() {
    return nextWarpzone() - 1;
}

export function getWarpzoneTicks() {
    return game.tick - storage.warp_zone_start_tick;
}

export function currentSurfaces() {
    return {
        factory: game.surfaces.nauvis,
        logistics: game.surfaces[surfaces.logistics],
        ground: storage.current_surface
    };
}

export function onInit() {
    storage.next_warp_zone = 1;
    storage.warp_zone_start_tick = 600;
}

export function setCurrentReactor(reactor: LuaEntity) {
    storage.reactor_destruction = script.register_on_object_destroyed(reactor)[0];
    storage.reactor = reactor;
}

export function getWarpReactor() {
    return storage.reactor;
}

export function isCurrentReactorDestroyed(event: OnObjectDestroyedEvent) {
    return event.registration_number == storage.reactor_destruction;
}
